### DISCUSSION TITLE
inspur.sm

### DISCUSSION THREAD

**Discussion by ISIB-Group** [ansible-collections/ansible-inclusion#8] [Resolved reviews]
Ansible Galaxy: https://galaxy.ansible.com/inspur/sm
GitHub repo: https://github.com/ISIB-Group/inspur.sm
GitHub repo: https://github.com/ISIB-Group/inspur.sm/issues
Automation Hub: https://cloud.redhat.com/ansible/automation-hub/repo/published/inspur/sm
Contacts: @ISIB-Group

I think our collection should meet the collection requirements of the collection.

**Comment by ISIB-Group** [comment]
Modify the repository name:
GitHub repo: https://github.com/ISIB-Group/inspur.sm
Issue tracker: https://github.com/ISIB-Group/inspur.sm/issues

**Comment by felixfontein** [comment]
There are some issues which must be addressed. The most serious is about check mode. All modules claim they support check mode, but nowhere `module.check_mode` is used - the modules seem to do the same action when run with or without check mode.

**Public availability and communication:**
- [x] published on Ansible Galaxy [Ansible Galaxy](https://galaxy.ansible.com) with version 1.0.0 or later
- [x] has a policy of releasing, versioning and deprecation announced to contributors and users in some way
  - Comment: I can't find any information on that in the collection repository.
- [x] have a Code of Conduct (CoC) compatible with the [Ansible Code of Conduct (CoC)](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)
  - Comment: I can't find any information on that in the collection repository.
- [x] have a publicly available issue tracker that does not require a paid level of service to create an account or view issues

**Standards and documentation:**
- [x] adheres to [semantic versioning](https://semver.org/)
- [x] follows licensing rules
- [x] follows the [Ansible documentation standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html) and the [style guide](https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#style-guide)
  - Acceptable, though improvements are possible. For example, in `DNS Server3 ipv4 or ipv6 address,Required when I(dns_manual=manual).` there should be a `. ` instead of `,` before `Required`. And the correct capitalization is `IPv4` and `IPv6` and not `ipv4` and `ipv6`.
- [x] follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html)
  - While the modules claim they support check mode, I do not see where this is actually implemented. It looks like the modules do the same action both with and without check mode!
- [x] has `README.md`
- [x] documentation, examples, and return use FQCNs for `M(..)`, examples, and `seealso` subsections
  - Comment: examples use `collections:` keyword and not FQCN
- [x] modules (or plugins) from ansible-base use `ansible.builtin.` as a FQCN prefix
- [x] documentation and return sections use `version_added:` containing the *collection* version for which an option, module or plugin was added
  - Comment: all content seems to have been added for 0.1.0, which was never released on Galaxy.
- [x] FQCNs is used for `extends_documentation_fragment:`, unless you are referring to doc_fragments from ansible-base

**Collection management:**
- [x] `galaxy.yml` having `tags` field set
- [x] collection dependencies must have a lower bound on the version which is at least 1.0.0, and are all part of the `ansible` package
- [x] `meta/runtime.yml` defines the minimal version of Ansible which the collection works with
  - Comment: There is no `meta/runtime.yml` file
- [x] has changelog, preferably with `changelogs/changelog.yaml`

**Tests:**
- [x] passed `ansible-test sanity`
- [x] if `test/sanity/ignore*.txt` exists, it does not contain error codes listed [here](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#ci-testing)
- [x] has CI tests up and running against each of the "major versions" of `ansible-base`/`ansible-core` that the collection supports
  - Comment: CI is only run for stable-2.10 and devel, but **not** for stable-2.9, which the collection supports according to documentation.
  - Comment: sanity tests specify `--python 3.6` to `ansible-test sanity --docker`; this **MUST** be removed so that all sanity tests run!
- [x] all CI tests run against every pull request
- [x] all CI tests run against a commit that releases the collection; if they don't pass, the collection won't be released
  - Comment: It is not clear to me whether this is the case.
- [x] all CI tests run regularly (nightly, or at least once per week)
  - Comment: This seems not to be the case.

**Reply by ISIB-Group** [reply]
> * [ ]  has a policy of releasing, versioning and deprecation announced to contributors and users in some way
>   * Comment: I can't find any information on that in the collection repository.

We have added Release Notes to[ README. md](https://github.com/ISIB-Group/inspur.sm/blob/main/README.md) so that we can publish relevant information through logs.

> * [ ]  have a Code of Conduct (CoC) compatible with the [Ansible Code of Conduct (CoC)](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)
>   * Comment: I can't find any information on that in the collection repository.

We have added Ansible Code of Conduct (COC) to [README. md](https://github.com/ISIB-Group/inspur.sm/blob/main/README.md)

> * [ ]  follows the [Ansible documentation standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html) and the [style guide](https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#style-guide)
>   
>   * Acceptable, though improvements are possible. For example, in `DNS Server3 ipv4 or ipv6 address,Required when I(dns_manual=manual).` there should be a `. ` instead of `,` before `Required`. And the correct capitalization is `IPv4` and `IPv6` and not `ipv4` and `ipv6`.

We have made modifications to this problem.

> * [ ]  follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html)
>   
>   * While the modules claim they support check mode, I do not see where this is actually implemented. It looks like the modules do the same action both with and without check mode!

Regarding check mode, is it to ensure that the system is not modified when running check mode?

> * [ ]  documentation, examples, and return use FQCNs for `M(..)`, examples, and `seealso` subsections
>   
>   * Comment: examples use `collections:` keyword and not FQCN

We have made modifications to this problem,please check if it is correct.

> **Collection management:**
> 
> * [ ]  `meta/runtime.yml` defines the minimal version of Ansible which the collection works with
>   
>   * Comment: There is no `meta/runtime.yml` file

Added [runtime.yml](https://github.com/ISIB-Group/inspur.sm/blob/main/meta/runtime.yml) file to support 2.10, and changed the supported Ansible version information in README.md to support 2.10 only.

> 
> **Tests:**
> 
> * [ ]  has CI tests up and running against each of the "major versions" of `ansible-base`/`ansible-core` that the collection supports
>   
>   * Comment: CI is only run for stable-2.10 and devel, but **not** for stable-2.9, which the collection supports according to documentation.
>   * Comment: sanity tests specify `--python 3.6` to `ansible-test sanity --docker`; this **MUST** be removed so that all sanity tests run!

Modify the [ansible-test.yml](https://github.com/ISIB-Group/inspur.sm/blob/main/.github/workflows/ansible-test.yml) file to add Python versions.Please check whether it meets the requirements.

> * [ ]  all CI tests run against a commit that releases the collection; if they don't pass, the collection won't be released
>   
>   * Comment: It is not clear to me whether this is the case.

Yes.if they don't pass, the collection won't be released,That's for sure.

> * [ ]  all CI tests run regularly (nightly, or at least once per week)
>   
>   * Comment: This seems not to be the case.

What can I do to make CI run regularly? Do I need to run it manually?

**Reply by ISIB-Group** [reply]
> * [ ]  follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html)
>   
>   * While the modules claim they support check mode, I do not see where this is actually implemented. It looks like the modules do the same action both with and without check mode!

I have tried to add a handling of the check mode in [module_utils/ism.py](https://github.com/ISIB-Group/inspur.sm/blob/main/plugins/module_utils/ism.py), please see if it works.

**Reply by felixfontein** [reply]
> > * [ ]   follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html)
> >   
> >   * While the modules claim they support check mode, I do not see where this is actually implemented. It looks like the modules do the same action both with and without check mode!
> 
> I have tried to add a handling of the check mode in [module_utils/ism.py](https://github.com/ISIB-Group/inspur.sm/blob/main/plugins/module_utils/ism.py), please see if it works.

It always returns `changed == True`. Is this always true for all your modules?

**Reply by ISIB-Group** [reply]
> > > * [ ]    follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html)
> > >   
> > >   * While the modules claim they support check mode, I do not see where this is actually implemented. It looks like the modules do the same action both with and without check mode!
> > 
> > 
> > I have tried to add a handling of the check mode in [module_utils/ism.py](https://github.com/ISIB-Group/inspur.sm/blob/main/plugins/module_utils/ism.py), please see if it works.
> 
> It always returns `changed == True`. Is this always true for all your modules?

Yes, I don't know if that's OK.

**Reply by ISIB-Group** [reply]
In check mode, it always returns changed == True;Instead of checking the pattern, it will go to the real implementation.

**Reply by felixfontein** [reply]
> > * [ ]   has a policy of releasing, versioning and deprecation announced to contributors and users in some way
> >   
> >   * Comment: I can't find any information on that in the collection repository.
> 
> We have added Release Notes to[ README. md](https://github.com/ISIB-Group/inspur.sm/blob/main/README.md) so that we can publish relevant information through logs.

A releasing/versioning/deprecation policy is more something saying that how often you plan to release (can be anything from "once per month" to "whenever we have new bugfixes and releases"), how often you release new major versions ("no new major release planned (since we don't want to remove anything yet)"), and how long things will be deprecated before removal ("if we want to remove something, we deprecate it at least 6 months before that").

> > * [ ]   follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html)
> >   
> >   * While the modules claim they support check mode, I do not see where this is actually implemented. It looks like the modules do the same action both with and without check mode!
> 
> Regarding check mode, is it to ensure that the system is not modified when running check mode?

Yes, and that the modules try to return the same results as when check mode is not specified. This is of course not always possible, but at least the `changed` state should come close to running without check mode. If that's too complicated to implement, you could simply say `supports_check_mode=False`.

> > * [ ]   documentation, examples, and return use FQCNs for `M(..)`, examples, and `seealso` subsections
> >   
> >   * Comment: examples use `collections:` keyword and not FQCN
> 
> We have made modifications to this problem,please check if it is correct.

Looks good!

> > **Collection management:**
> > 
> > * [ ]   `meta/runtime.yml` defines the minimal version of Ansible which the collection works with
> >   
> >   * Comment: There is no `meta/runtime.yml` file
> 
> Added [runtime.yml](https://github.com/ISIB-Group/inspur.sm/blob/main/meta/runtime.yml) file to support 2.10, and changed the supported Ansible version information in README.md to support 2.10 only.

The README still mentions 2.9.10, you probably want to replace that by 2.10.0.

> > **Tests:**
> > 
> > * [ ]   has CI tests up and running against each of the "major versions" of `ansible-base`/`ansible-core` that the collection supports
> >   
> >   * Comment: CI is only run for stable-2.10 and devel, but **not** for stable-2.9, which the collection supports according to documentation.
> >   * Comment: sanity tests specify `--python 3.6` to `ansible-test sanity --docker`; this **MUST** be removed so that all sanity tests run!
> 
> Modify the [ansible-test.yml](https://github.com/ISIB-Group/inspur.sm/blob/main/.github/workflows/ansible-test.yml) file to add Python versions.Please check whether it meets the requirements.

It does run all sanity tests now for all Python versions, though multiple times. You can remove all entries from the `python` matrix except one (I suggest 3.8), since you no longer add `--python` to `ansible-test sanity --docker` you can limit to one Python version for setting up Ansible.

> > * [ ]   all CI tests run against a commit that releases the collection; if they don't pass, the collection won't be released
> >   
> >   * Comment: It is not clear to me whether this is the case.
> 
> Yes.if they don't pass, the collection won't be released,That's for sure.

:+1:

> > * [ ]   all CI tests run regularly (nightly, or at least once per week)
> >   
> >   * Comment: This seems not to be the case.
> 
> What can I do to make CI run regularly? Do I need to run it manually?

See here: https://github.com/ansible-collections/collection_template/blob/main/.github/workflows/ansible-test.yml#L12-L15

**Reply by ISIB-Group** [reply]
> * [ ]  has a policy of releasing, versioning and deprecation announced to contributors and users in some way
>   
>   * Comment: I can't find any information on that in the collection repository.

We developed the release notes and included the roadmap in README.md.

> * [ ]  follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html)
>   
>   * While the modules claim they support check mode, I do not see where this is actually implemented. It looks like the modules do the same action both with and without check mode!

We will change the supports_check_mode value to False in all modules, and we will gradually support checking mode later.

> * [ ]  all CI tests run regularly (nightly, or at least once per week)
>   
>   * Comment: This seems not to be the case.

We have added a script that is timed to execute in [ansible-test.yml](https://github.com/ISIB-Group/inspur.sm/blob/main/.github/workflows/ansible-test.yml)

**Reply by dmsimard** [reply]
> What can I do to make CI run regularly? Do I need to run it manually?

I wanted to mention that scheduled runs were added in https://github.com/ISIB-Group/inspur.sm/commit/17ef56536910893580f16631c859264a51cf695d and have been running successfully:
![Screenshot from 2021-01-19 10-31-53](https://user-images.githubusercontent.com/1291204/105055950-a1de3300-5a41-11eb-89f8-073532b05fe8.png)

**Comment by tadeboro** [comment]
Hello.

The community has agreed on a list of criteria that must be met to move forward with the inclusion of a collection in the Ansible package. For this discussion, we will use a shortened checklist to verify and record compliance, but you can find the full details in the [Collection Requirements](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst) documentation.

This is a new and gradual process, and we will edit this comment as we progress along. The items in the checklist can be in one of the following states:

  * Items with a checkmark and no attached comment are OK and need no intervention from your side.
  * Items with a checkmark that contain a comment will not block the inclusion process, but by following the advice given in the comment, the collection's quality can be improved.
  * Items with no checkmark and no extra details are waiting for evaluation.
  * Items with no checkmark that contain at least one comment need your attention. Issues in comments marked as *required* need to be fixed before collection can be included in Ansible. Comments marked as *optional* are just friendly advice and will not block the inclusion.

Feel free to ask any questions and provide feedback. Thanks in advance for your understanding.

**Public availability and communication:**
- [x] published on Ansible Galaxy [Ansible Galaxy](https://galaxy.ansible.com) with version 1.0.0 or later
- [x] has a policy of releasing, versioning and deprecation announced to contributors and users in some way
- [x] have a Code of Conduct (CoC) compatible with the [Ansible Code of Conduct (CoC)](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)
- [x] have a publicly available issue tracker that does not require a paid level of service to create an account or view issues

**Standards and documentation:**
- [x] adheres to [semantic versioning](https://semver.org/)
- [x] follows licensing rules
- [x] follows the [Ansible documentation standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html) and the [style guide](https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#style-guide)
- [x] follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html)
- [x] has `README.md`
- [x] documentation, examples, and return use FQCNs for `M(..)`, examples, and `seealso` subsections
- [x] modules (or plugins) from ansible-base use `ansible.builtin.` as a FQCN prefix
- [x] documentation and return sections use `version_added:` containing the *collection* version for which an option, module, or plugin was added
- [x] FQCNs is used for `extends_documentation_fragment:`, unless you are referring to doc_fragments from ansible-base

**Collection management:**
- [x] `galaxy.yml` having `tags` field set
- [x] collection dependencies must have a lower bound on the version which is at least 1.0.0 and are all part of the `ansible` package
- [x] `meta/runtime.yml` defines the minimal version of Ansible which the collection works with
- [x] has changelog, preferably with `changelogs/changelog.yaml`

**Tests:**
- [x] passed `ansible-test sanity`
- [x] if `test/sanity/ignore*.txt` exists, it does not contain error codes listed [here](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#ci-testing)
- [x] has CI tests up and running against each of the "major versions" of `ansible-base`/`ansible-core` that the collection supports
- [x] all CI tests run against every pull request
- [x] all CI tests run against a commit that releases the collection; if they don't pass, the collection won't be released
- [x] all CI tests run regularly (nightly, or at least once per week)

**Reply by tadeboro** [reply]
I did a review of this collection and found a few issues. Most of them should be relatively straightforward to resolve, but the one about following development conventions is a bit trickier. The fact that the `inspur.sm` collection is a thin wrapper around the underlying SDK makes writing playbooks that enforce the desired state a real chore. Of course, I cannot be sure this is the case because I do not know what SDK is doing (I could not find it anywhere), but the presence of add_ and edit_ modules give a hint about what would happen if I run the playbook that, for example, adds an LDAP user, twice In a row.

**Reply by ISIB-Group** [reply]
> I did a review of this collection and found a few issues. Most of them should be relatively straightforward to resolve, but the one about following development conventions is a bit trickier. The fact that the `inspur.sm` collection is a thin wrapper around the underlying SDK makes writing playbooks that enforce the desired state a real chore. Of course, I cannot be sure this is the case because I do not know what SDK is doing (I could not find it anywhere), but the presence of add_ and edit_ modules give a hint about what would happen if I run the playbook that, for example, adds an LDAP user, twice In a row.

The inspur_sm_sdk is a functional package of the administration tool for managing Inspur servers, providing an interface for managing Inspur servers.But he has not released it to the public.And SDK is only supported for a few specific models of Inspur, so the test needs to have specific managed nodes.

**Reply by ISIB-Group** [reply]
> * [ ]  have a Code of Conduct (CoC) compatible with the [Ansible Code of Conduct (CoC)](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)
>   **Required:** README file only contains a link to Ansible CoC, but It does not specify that this project follows that CoC. I know the official template contains the same information in README file, but the template repository also contains a CODE_OF_CONDUCT.md which makes things more clear. There are two ways forward here: either copy the CODE_OF_CONDUCT.md from the template repo or clearly state in README file what CoC is in use when interacting with the inspur.sm community.

We added the [CODE_OF_CONDUCT.md](https://github.com/ansible-collections/collection_template/blob/main/CODE_OF_CONDUCT.md).

> * [ ]  follows the [Ansible documentation standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html) and the [style guide](https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#style-guide)
>   **Required:** Ansible has no concept of master that is described in the README file. Proper nomenclature is control node and managed nodes or hosts.

We have modified the description of Ansible and the control node in the [README.md](https://github.com/ISIB-Group/inspur.sm/blob/main/README.md).

> * [ ]  follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html)
>   **Required:** The main issue I have with this collection is that its content is a very thin wrapper around the python SDK. Because of this, we have quadruples of `add_*`, `edit_*`, `del_*` and `*_info` modules that execute CRUD operations on target. Also, the python SDK seems to be proprietary, which makes it really hard to predict how modules behave when used. For example, I still do not know when to expect changed to be true or false.

Regarding the SDK, even with this SDK, the program can't really run because the SDK needs to connect and manage the Inspur server to run.How should we deal with this situation?

> * [ ]  all CI tests run against a commit that releases the collection; if they don't pass, the collection won't be released
>   **Required:** CI is not triggered on pushes, which means that CI does not see all commits that land in the default branch. This can cause collection to be released based on a commit that did not go through the CI.

We add a script about push in the [ansible-test.yml](https://github.com/ISIB-Group/inspur.sm/blob/main/.github/workflows/ansible-test.yml).

**Reply by dmsimard** [reply]
Hello,

First I wanted to thank you for your interest and for being responsive to improvement opportunities so far.

I wanted to add to this comment from @tadeboro:

> I did a review of this collection and found a few issues. Most of them should be relatively straightforward to resolve, but the one about following development conventions is a bit trickier. The fact that the `inspur.sm` collection is a thin wrapper around the underlying SDK makes writing playbooks that enforce the desired state a real chore. Of course, I cannot be sure this is the case because I do not know what SDK is doing (I could not find it anywhere), but the presence of add_ and edit_ modules give a hint about what would happen if I run the playbook that, for example, adds an LDAP user, twice In a row.

I agree with the sentiment here and wanted to draw a comparison with other Ansible modules as an example for the sake of discussion.

The [user](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/user_module.html) module provides the ability to create a user (``state: present``), edit it (for example changing the shell) or delete it (``state: absent``). This is abstracted by the module which handles idempotency and takes care of running the right commands with the right arguments.

On the other hand, the modules in the collection follow a trend of ``add_``, ``edit_`` and ``del_`` like:
- add_ldap_group
- edit_ldap_group
- del_ldap_group

This is somewhat of an anti-pattern and I don't know to what extent it could be addressed without a significant refactor for single modules to support the three operations with idempotency.

Is this something you agree with and would be willing to consider doing in the future ?

Edit: for clarity, this is best explained in the [best practices](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html#following-ansible-conventions):
> ## Following Ansible conventions
> [...]
> - Implement declarative operations (not CRUD) so the user can ignore existing state and focus on final state. For example, use ``started/stopped``, ``present/absent``.

**Reply by tadeboro** [reply]
Did look over the https://github.com/ISIB-Group/inspur.sm/pull/30 and once this is merged, the CoC, documentation standards, and CI section will get a checkmark.

**Reply by tadeboro** [reply]
> Regarding the SDK, even with this SDK, the program can't really run because the SDK needs to connect and manage the Inspur server to run.How should we deal with this situation?

Not being able to run the modules is not problematic here. Not being able to inspect what the SDK does is. This means that there is no way a potential contributor can contribute to this collection without having some kind of contract with the SDK vendor.

**Reply by ISIB-Group** [reply]
> The [user](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/user_module.html) module provides the ability to create a user (`state: present`), edit it (for example changing the shell) or delete it (`state: absent`). This is abstracted by the module which handles idempotency and takes care of running the right commands with the right arguments.
> 

Thank you, I understand what you said, I will revise it, but it may take two days.

**Reply by dmsimard** [reply]
@ISIB-Group I see you added ``ad_group``, ``ldap_group``, ``user`` and ``user_group`` modules in https://github.com/ISIB-Group/inspur.sm/commit/1fdd6301b5c77efa8eb37132a71170524280404e which seem like an improvement at first glance. Do you plan to deprecate the previous modules in favor of these new ones ?

**Reply by tadeboro** [reply]
I rechecked the collection, and as predicted, the easy fixes were indeed implemented. But the main issue with collection modules being a paper-thin wrapper around underlying SDK is still present. To be sure, will the following task make sure the user is gone from the BMC:

```
- name: Make sure user account is not present on BMC
  inspur.sm.user:
    uname: some_user
    state: absent
```

Looking at the code, I am inclined to say no, but I would love to be proven wrong.

**Reply by ISIB-Group** [reply]
> I rechecked the collection, and as predicted, the easy fixes were indeed implemented. But the main issue with collection modules being a paper-thin wrapper around underlying SDK is still present. To be sure, will the following task make sure the user is gone from the BMC:
> 
> ```
> - name: Make sure user account is not present on BMC
>   inspur.sm.user:
>     uname: some_user
>     state: absent
> ```
> 
> Looking at the code, I am inclined to say no, but I would love to be proven wrong.

This can work, and here's what I did.But you should also add information about the logged-in user.

[root@localhost ~]# ansible localhost -m inspur.sm.user -a 'host=10.48.3.53 username=admin password=Inspur1! state=present uname=wbs upass=12345678 role_id=Administrator priv=kvm,vmm'
localhost | CHANGED => {
    "Message": [
        "add user success."
    ],
    "State": "Success",
    "changed": true
}
[root@localhost ~]# ansible localhost -m inspur.sm.user -a 'host=10.48.3.53 username=admin password=Inspur1! state=absent uname=wbs'
localhost | CHANGED => {
    "Message": [
        "del user success."
    ],
    "State": "Success",
    "changed": true
}

**Reply by ISIB-Group** [reply]
> @ISIB-Group I see you added `ad_group`, `ldap_group`, `user` and `user_group` modules in [ISIB-Group/inspur.sm@1fdd630](https://github.com/ISIB-Group/inspur.sm/commit/1fdd6301b5c77efa8eb37132a71170524280404e) which seem like an improvement at first glance. Do you plan to deprecate the previous modules in favor of these new ones ?

Yes, we plan to deprecate the modules replaced by 'ad_group', 'ldap_group', 'user' and 'user_group', but we haven't done that yet.

**Comment by felixfontein** [comment]
About the inspur_sm_sdk: under which license is it distributed? Since the collection is GPLv3, the SDK needs to be distributed under a compatible license (see for example https://www.gnu.org/licenses/gpl-faq.html#MoneyGuzzlerInc). That would entitle users whom you gave it to to simply upload it to the web for anyone else. If that's the case, why not publish it anyway? :)

**Reply by ISIB-Group** [reply]
I have seen the relevant GPLv3 content, but there is also a reference in our JDK that is not open source and I will need two days to modify it and then release it.

**Reply by ISIB-Group** [reply]
> About the inspur_sm_sdk: under which license is it distributed? Since the collection is GPLv3, the SDK needs to be distributed under a compatible license (see for example https://www.gnu.org/licenses/gpl-faq.html#MoneyGuzzlerInc). That would entitle users whom you gave it to to simply upload it to the web for anyone else. If that's the case, why not publish it anyway? :)

We have published the SDK and point to its URL in [README.md](https://github.com/ISIB-Group/inspur.sm/blob/main/README.md).

**Reply by dmsimard** [reply]
@ISIB-Group thank you for publishing the SDK! It will make it easier for users to understand and contribute.

**Reply by gundalow** [reply]
Thank you for publishing the SDK (under GPLv3). 
As this discussion is resolved I'll hide this discussion.

**Comment by abadger** [comment]
Just a note.  The deadline for new collections to be accepted was today, 27-01-2021, but we voted at today's community irc meeting to extend that to Friday:

* inspur.sm has until Friday to be ready for another review.
* Reviewers can approve or say not ready by 2-2.
* If it's not approved by 2-2, it has to go to ansible-4 instead.

**Reply by ISIB-Group** [reply]
Thanks for your reminding. Now inspur.sm has met all the requirements of Ansible, so we are ready.
Are there any new requirements in the next audit, or what should be paid attention to?Thanks again.

**Reply by felixfontein** [reply]
Thanks for the 1.1.0 release, I'll try to take a look today, but maybe I only manage over the weekend.

**Reply by ISIB-Group** [reply]
Thank you. Is there anything else I need to do?

**Reply by tadeboro** [reply]
I did a fresh review and found just one minor thing that needs to be fixed: `version_added` fields on new modules should be set to 1.1.0, not to 1.0.0. Once this is fixed and 1.1.1 with fixes is released, this collection is ready for inclusion as far as I am concerned.

(Since I was already looking over the code, I prepared https://github.com/ISIB-Group/inspur.sm/pull/35 to help out a bit.)

**Reply by ISIB-Group** [reply]
Thank you very much for your support, and we have now released version 1.1.1

**Reply by felixfontein** [reply]
I looked again at it, and I think I found one aspect which we overlooked before: the `_info` modules should have `supports_check_mode=True`, since they do not modify anything (see https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html#following-ansible-conventions: "Implement check_mode in all *_info and *_facts modules.").

Since these modules do not modify anything (if they do, it's a bug!), setting `supports_check_mode=True` for them should be no problem.

**Reply by ISIB-Group** [reply]
We have changed all modules ending in _info from 'supports_check_mode=False' to 'supports_check_mode=True'.The _info modules never change anything, because they are all implementation fetches.

**Reply by felixfontein** [reply]
Awesome! Looks good now, thanks a lot! :)

**Reply by tadeboro** [reply]
Thank you all for either work on the collection itself or for reviewing the submission. I created a PR to have the inspur.sm Ansible Collection included in Ansible 3.0.0 (https://github.com/ansible-community/ansible-build-data/pull/60).

**Reply by ISIB-Group** [reply]
Thank you all for your help and support for our collection. Thank you very much.

**Comment by abadger** [comment]
Thanks for all the hard work getting this ready for inclusion!  We look forward to working with you into the future :-)
